https-bastion.md

```diff
@@ -0,0 +1,137 @@
+most HTTP servers.
+
+Backends are authenticated by their Ed25519 key, and the hash of that public key
+becomes the first segment of the path of client requests. To prevent abuse, it
+is RECOMMENDED that bastions apply allowlists for the backends that they serve.
+
+Like regular HTTPS reverse proxies, a bastion is in the position of blocking,
+observing, and manipulating the backend’s traffic. Bastions SHOULD only be used
+for self-authenticating protocols that don't rely exclusively on TLS for
+authentication or confidentiality. Alternatively, backends MAY use bastions they
+trust as part of their broad hosting infrastructure.
+
+Bastions were designed to enable synchronous [transparency log witnesses][]
+which prefer not to accept connections directly from the Internet.
+
+## API
+
+At a high level, backends connect to the bastion authenticating with their
+Ed25519 key and start serving HTTP/2 on that connection. The bastion then
+proxies to that connection all requests that include the backend's public key
+hash in the URL path.
+
+### Backend to bastion connection
+
+A backend connects to the bastion’s backend endpoint with TLS 1.3, specifying
+the ALPN protocol `bastion/0`. Other TLS versions MUST be rejected, as only TLS
+1.3 hides client certificates from network observers. The backend MUST present as
+its client certificate a self-signed Ed25519 certificate containing the backend
+public key. The backend verifies the bastion’s TLS certificate chain as usual.
+The bastion checks the backend public key against an allowlist.
+
+After opening the connection, the backend starts serving HTTP/2 traffic on it as
+if it was a client-initiated connection. HTTP/2’s multiplexing allows serving
+multiple parallel requests on a single connection efficiently. None of the
+backend's HTTP APIs are modified, except for observing the `X-Forwarded-For`
+header.
+
+Appendix B presents an example Go adapter that turns a regular HTTP server into
+a bastion backend.
+
+### Client to bastion requests
+
+The bastion accepts HTTP requests at
+
+```
+https://<bastion host>/<key hash>/<path>
+```
+
+where "key hash" MUST be a lowercase hex-encoded SHA-256 hash of a 32-byte
+Ed25519 public key.
+
+If the backend key doesn’t correspond to any open backend connection, the
+bastion MUST serve a 502 Bad Gateway response.
+
+Otherwise, the bastion MUST remove _all_ `X-Forwarded-For` headers from the
+request, add a single `X-Forwarded-For` header with the IP address of the
+client, and proxy the request as `/<path>` over the backend connection. Note
+that the `/<key hash>` portion of the path is trimmed.
+
+The bastion MUST ignore all caching headers in both request and response.
+
+## Appendix A — Security analysis of reusing a witness key
+
+Transparency log witnesses MAY choose to reuse the witness key as their bastion
+backend key.
+
+If domain separation between the TLS handshake signatures, the X.509 certificate
```

Comment:
This section might be more readable if we re-flow the "arc" a bit, e.g.:
 
```
For this to be done safely, there must be domain separation between all protocols the key is used in to mitigate the risk of cross-protocol attacks.

This requirement holds true for the witness key use as described in this document since the various signed message types have no common prefix, guaranteeing domain separation:
* TLS 1.3 ...
* X.509 ...
* The witness protocol ...
```

There's a hanging implication that if you're using the witness key for yet something else, you should make sure that this domain separation holds there too - don't know if we should explicitly call that out or leave it as "obvious" and/or out of scope?
(I don't feel strongly either way, although perhaps being explicit about this here may save someone from a painful mis-step?)

Reply:
Much better, thanks!

Added a NOT RECOMMENDED / MUST be analyzed for further reuse.